sormas-ui/src/main/java/de/symeda/sormas/ui/externalmessage/labmessage/processing/AbstractLabMessageProcessingFlow.java

```diff
@@ -109,119 +137,196 @@ private FlowThen<SampleAndPathogenTests> doCreateEventParticipantFlow(
+								createSampleForEventParticipant,
+								flow,
+								labMessage).getResult());
+					})
 					.when(
 						EventValidationResult::isEventSelectionCanceled,
 						(vf, v) -> vf.then(ignored -> doCreateEventParticipantFlow(flow, labMessage).getResult()))
 					.then(s -> ProcessingResult.continueWith(s.getData()).asCompletedFuture()))
 			.then(s -> ProcessingResult.continueWith(s.getData()).asCompletedFuture());
 	}
 
-	private FlowThen<SampleAndPathogenTests> doCreateEventParticipantAndSampleFlow(
+	private FlowThen<RelatedSamplesReportsAndPathogenTests> doCreateEventParticipantAndSamplesFlow(
 		EventDto event,
 		PersonDto person,
 		FlowThen<?> flow,
 		ExternalMessageDto labMessage) {
 
+		//@formatter:off
 		return flow.then(ignored -> createEventParticipant(event, person, labMessage))
-			.then(ep -> createSampleAndPathogenTests(ep.getData().getEventParticipant().toReference(), ep.getData().getEvent(), labMessage, true));
+				.then(p -> {
+					IntFunction<CompletionStage<ProcessingResult<SampleAndPathogenTests>>> createSampleForEventParticipant =
+						sampleReportIndex -> createOneSampleAndPathogenTests(
+							p.getData().eventParticipant.toReference(), 
+							p.getData().getEvent(),
+							labMessage,
+							sampleReportIndex,
+							true);
+
+					return doPickOrCreateSamplesFlow(
+						c -> c.sampleCriteria(new SampleCriteria().eventParticipant(p.getData().getEventParticipant().toReference())),
+						createSampleForEventParticipant,
+						flow,
+						labMessage).getResult();
+		});
+		//@formatter:on
 	}
 
-	private FlowThen<SampleAndPathogenTests> doCaseSelectedFlow(
-		CaseSelectionDto caseSelection,
-		FlowThen<PersonAndPickOrCreateEntryResult> flow,
+	private FlowThen<RelatedSamplesReportsAndPathogenTests> doPickOrCreateSamplesFlow(
+		Consumer<SampleSimilarityCriteria> addSampleSearchCriteria,
+		IntFunction<CompletionStage<ProcessingResult<SampleAndPathogenTests>>> createSampleAndPathogenTests,
+		FlowThen<?> flow,
 		ExternalMessageDto labMessage) {
 
-		CaseDataDto caze = FacadeProvider.getCaseFacade().getCaseDataByUuid(caseSelection.getUuid());
+		List<SampleReportDto> sampleReports = labMessage.getSampleReportsNullSave();
+		if (sampleReports.size() > 1) {
+			flow = flow.then(
+				result -> handleMultipleSampleConfirmation().thenCompose(
+					next -> ProcessingResult
+						.<Void> withStatus(Boolean.TRUE.equals(next) ? ProcessingResultStatus.CONTINUE : ProcessingResultStatus.CANCELED)
+						.asCompletedFuture()));
+		}
+
+		RelatedSamplesReportsAndPathogenTests allSamplesAndPathogenTests = new RelatedSamplesReportsAndPathogenTests();
+		FlowThen<SampleAndPathogenTests> currentFlow = null;
+		int i = 0;
+		do {
+			int currentSampleReportIndex = i;
+			currentFlow = doSinglePickOrCreateSampleFlow(
+				addSampleSearchCriteria,
+				createSampleAndPathogenTests,
+				currentFlow == null ? flow : currentFlow,
```

Comment:
You can initialize `currentFlow` to be `flow` and than the ternary will not be needed